keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,810 @@
+      # in the version specified by subjectAccessReviewVersion in the request variable.
+      # Documentation on CEL: https://kubernetes.io/docs/reference/using-api/cel/
+      #
+      # only send resource requests to the webhook
+      - expression: has(request.resourceAttributes)
+      # only intercept requests to kube-system
+      - expression: request.resourceAttributes.namespace == 'kube-system'
+      # don't intercept requests from kube-system service accounts
+      - expression: !('system:serviceaccounts:kube-system' in request.user.groups)
+  - type: Node
+  - type: RBAC
+  - type: Webhook
+    webhook:
+      name: in-cluster-authorizer
+      authorizedTTL: 5m
+      unauthorizedTTL: 30s
+      timeout: 3s
+      subjectAccessReviewVersion: v1
+      failurePolicy: NoOpinion
+      connectionInfo:
+        type: InClusterConfig
+```
+
+Validation will allow multiple authorizers of type "Webhook" to be added to the
+config, but at most one authorizer each for other types. The ordering of this chain will
+be decided by the order specified in the file.
+
+The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and
+`subjectAccessReviewVersion` accept values corresponding to flags
+`--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`,
+`--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version`
+respectively.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
+flag is not supplied. While configuring authorization modes using the file config,
+the version supported by a webhook has to be mentioned using a required field
+`subjectAccessReviewVersion`.
+
+The user can define a CEL expression to determine whether a request needs to dispatched
+to the authz webhook for which the expression has been defined. The user would have access
+to a `request` variable containing a `SubjectAccessReview` object in the version specified
+by `subjectAccessReviewVersion`.
+
+When no matchConditions are satisfied for a request, the webhook would be skipped. In such
+situations, the decision is logged in the audit log with the `webhookskipped.k8s.io` annotation.
+Benefit of this is that resource and user info will also be logged.
+
+The code path for enabling the above will only be triggered if the feature flag is enabled
+while the feature is in alpha and beta.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_decisions_total`
+
+This will be incremented on round-trip of an authorizer. It will track total
+authorization decision invocations across the following labels.
+
+Labels {along with possible values}:
+- `mode` {<authorizer_name>} # when authorizer is a webhook, prepend `webhook_`
+- `decision` {Allow, Deny}
+
+**Note:** Some examples of <authorizer_name>: `RBAC`, `Node`, `ABAC`, `webhook{,_<name>}`.
+If there is only one webhook and no name specified, there would be no `_<name>` suffix.
```

Comment:
Currently the `name` field is only required when there are multiple webhooks. What this is saying is if there's only one webhook and the name is NOT specified, then there's no `_<name>` suffix just `webhook`.  But otherwise, whenever `name` is specified, it will be added as a suffix `webhook_<name>`.

Reply:
I think it is far simpler to just make the name required.

Reply:
> I think it is far simpler to just make the name required.

Yeah this was basically what I was saying.

Reply:
After thinking through this, made `name` a require field.